ptodo.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man 5.0102 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\" ========================================================================
.\"
.IX Title "PTODO 1"
.TH PTODO 1 2025-01-23 "perl v5.40.0" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
todo \- a simple todo manager
.SH SYNOPSIS
.IX Header "SYNOPSIS"
.Vb 1
\&    todo [OPTIONS] DIRECTORY
\&
\&    Options:
\&        \-h, \-\-help       show this help message
\&        \-l, \-\-list       show the TODOs without updating TODO.md
\&        \-v, \-\-verbose    show each path before processing them
.Ve
.SH DESCRIPTION
.IX Header "DESCRIPTION"
\&\fItodo\fR is a simple perl script that help managing your TODO in a project.
.PP
It takes a directory to analyze as argument and will generate a TODO.md at the root of that directory with the found TODO.
.PP
If a \fITODO.md\fR is already in the root of the directory, \fItodo\fR will update that file, eventually marking some TODO as completed if they are not found in any sub path.
.SH "PARENTLESS TODOs"
.IX Header "PARENTLESS TODOs"
TODOs with no filepath as parent will not be tracked, they require to be automatically added and marked as
done in TODO.md.
.SH PROCESS
.IX Header "PROCESS"
.IP \(bu 4
Check if there is a \fITODO.md\fR in the target directory. If yes, load it as a hash.
.IP \(bu 4
Loop recursively in the filesystem to find all file that are readable and less than 10Mb in size.
.Sp
When finding one, will loop through the lines of the file and capture all TODOs which can be in those formats:
.Sp
.Vb 5
\&    TODO something to do
\&    TODO  something to do
\&    TODO: something to do
\&    TODO:  something todo
\&    ...
.Ve
.Sp
Basically any amount of \fI:\fR or space can be added after the word \fITODO\fR.
.Sp
The TODO line \fIMUST\fR starts by a "comment" such as '#' or '//' or ';' or '\-\-':
.Sp
.Vb 5
\&    # TODO something to do   (script langs)
\&    // TODO  something to do (C like)
\&    ;  TODO: something to do (asm)
\&    \-\- TODO:  something todo (lua)
\&    ...
.Ve
.IP \(bu 4
Will loop through the TODOs in the \fITODO.md\fR if any, to see which TODOs to mark as done (if they are not found in the files we check earlier).
.Sp
But skip the parentless TODOs are those one cannot be tracked (no file to check).
.IP \(bu 4
Finally, the todo hash will be converted into a markdown TODO list and depending on the given options will update the \fITODO.md\fR or print it to STDOUT.
.Sp
By default \fItodo\fR will update \fITODO.md\fR and only show pending tasks to STDOUT.
.SH IGNORE
.IX Header "IGNORE"
If you add a \fI.todoignore\fR file at the root of the directory to analyse, \fItodo\fR will skip matching path(s).
.SH AUTHOR
.IX Header "AUTHOR"
4zv4l \fI4zv4l@protonmail.com\fR


ptodo.pl
#!/usr/bin/env perl

use Getopt::Long;
use Pod::Usage;
use File::Find;
use File::Spec::Functions;
use v5.38;

# converts a hash to todo file content
# ------------------------------------
sub hash_to_todo(%hash) {
    my $result;
    # first dump the parentless todo
    for my $todo_hash ($hash{todo}->@*) {
        my ($todo, $status) = $todo_hash->@{'todo', 'status'}; 
        $result .= sprintf "- [%s] %s\n", ($status == 1 ? 'X' : ' '), $todo;
    }
    # then dump the todos with filepath as parent
    delete $hash{todo};
    for my $file_path (keys %hash) {
        $result .= "$file_path:\n";
        for my $todo_hash ($hash{$file_path}->@*) {
            my ($todo, $status) = $todo_hash->@{'todo', 'status'}; 
            $result .= sprintf "    - [%s] %s\n", ($status == 1 ? 'X' : ' '), $todo;
        }
    }
    return $result;
}

# converts a todo file content to hash
# ------------------------------------
sub todo_to_hash($todo) {
    my %result;
    my $current_file;
    my @lines = split '\n', $todo;
    for (@lines) {
        # if its a filepath
        if (not /^- / and /^(.+):$/) {
            $current_file = $1;
            $result{$current_file} = ();
        # if its a parentless todo (without indentation)
        } elsif (/^- \[(?<done>[Xx ])\] (?<todo>.+)/) {
            push @{$result{todo}}, { todo => $+{todo}, status => $+{done} eq ' ' ? 0 : 1};
        # if its a todo with a parent filepath (has indentation)
        } elsif (/^\s+- \[(?<done>[Xx ])\] (?<todo>.+)/) {
            push @{$result{$current_file}}, { todo => $+{todo}, status => $+{done} eq ' ' ? 0 : 1 };
        }
    }
    return %result;
}

# quick and dirty replacements for Path::Tiny
# -------------------------------------------
sub slurp($filepath) {
    open my $file, '<', $filepath or die "open(): $!";
    wantarray ? map {chomp; $_} (<$file>) : join '', <$file>;
}
sub spit($filepath, $data) {
    open my $file, '>', $filepath or die "open(): $!";
    print $file $data;
}

# parsing arguments
# -----------------
pod2usage(0) if @ARGV == 0;
Getopt::Long::Configure ("bundling");
GetOptions(
    "l|list" => \my $list,
    "u|undone" => \my $undone,
    "v|verbose" => \my $verbose,
    "h|help" => sub { pod2usage(0) },
) or pod2usage(1);
pod2usage({-message => "error: missing DIRECTORY", -exitval => 1}) if @ARGV == 0;
my $directory_path = $ARGV[0] =~ s/^\.\///r; # remove the optional ./ from the directory path

# check if directory exists
# -------------------------
die "$directory_path: No such readable directory\n" unless -d $directory_path and -e $directory_path; 

# load TODO.md and .todoignore if exists
# --------------------------------------
my $todo_path = catfile($directory_path, "TODO.md");
my $todoignore_path = catfile($directory_path, ".todoignore");
my %old_todo = (-e $todo_path and -r $todo_path) ? todo_to_hash(scalar slurp($todo_path)) : ();
my $todoignore = (-e $todoignore_path and -r $todoignore_path) ? join '|', slurp($todoignore_path) : undef;
my %todo;

# loop through files to add the TODOs
# -----------------------------------
find(sub {
    return unless -f and -r and (-s) < (10 * 1024 * 1024); # skip 10M+ files
    my $filepath = $File::Find::name;
    say "checking $filepath" if $verbose;
    if ($todoignore and $filepath =~ /$todoignore/) {
        say "skipped because matches todoignore pattern" if $verbose;
        return;
    }
    my @todos = map { /TODO[: ]+(.+)$/; {todo => $1, status => 0} } grep { /^(#|--|\/\/|;)(\s+)?TODO[: ]+(.+)$/ } (slurp($_));
    $todo{$filepath} = [] if @todos;
    for my $todo (@todos) {
        push @{$todo{$filepath}}, $todo;
    }
}, $directory_path);

# diff between maybe existing TODO.md and in memory TODOs
# -------------------------------------------------------
push $todo{todo}->@*, $old_todo{todo}->@* if $old_todo{todo};
delete $old_todo{todo};
for my $filepath (keys %old_todo) {
    for my $old_todo ($old_todo{$filepath}->@*) {
        unless (grep { $old_todo->{todo} eq $_->{todo} } $todo{$filepath}->@*) {
            push $todo{$filepath}->@*, {todo => $old_todo->{todo}, status => 1};
        }
    }
}

# dump the todo to TODO.md and show undone task to STDOUT
# -------------------------------------------------------
my $todo_md = hash_to_todo(%todo);
if ($list) {
    print $todo_md if $list;
} else {
    spit($todo_path, $todo_md);
    my @undone = grep { $_->{status} == 0 } $todo{todo}->@*;
    print map { "- [ ] $_->{todo}\n" } @undone;
    delete $todo{todo};
    for my $filepath (keys %todo) {
        my @undone = grep { $_->{status} == 0 } $todo{$filepath}->@*;
        say "$filepath:" if @undone;
        print map { "    - [ ] $_->{todo}\n" } @undone;
    }
}

__END__

=head1 NAME

todo - a simple todo manager

=head1 SYNOPSIS

    todo [OPTIONS] DIRECTORY

    Options:
        -h, --help       show this help message
        -l, --list       show the TODOs without updating TODO.md
        -v, --verbose    show each path before processing them

=head1 DESCRIPTION

I<todo> is a simple perl script that help managing your TODO in a project.

It takes a directory to analyze as argument and will generate a TODO.md at the root of that directory with the found TODO.

If a I<TODO.md> is already in the root of the directory, I<todo> will update that file, eventually marking some TODO as completed if they are not found in any sub path.

=head1 PARENTLESS TODOs

TODOs with no filepath as parent will not be tracked, they require to be automatically added and marked as
done in TODO.md.

=head1 PROCESS

=over

=item *

Check if there is a I<TODO.md> in the target directory. If yes, load it as a hash.

=item *

Loop recursively in the filesystem to find all file that are readable and less than 10Mb in size.

When finding one, will loop through the lines of the file and capture all TODOs which can be in those formats:

    TODO something to do
    TODO  something to do
    TODO: something to do
    TODO:  something todo
    ...

Basically any amount of I<:> or space can be added after the word I<TODO>.

The TODO line I<MUST> starts by a "comment" such as '#' or '//' or ';' or '--':

    # TODO something to do   (script langs)
    // TODO  something to do (C like)
    ;  TODO: something to do (asm)
    -- TODO:  something todo (lua)
    ...

=item *

Will loop through the TODOs in the I<TODO.md> if any, to see which TODOs to mark as done (if they are not found in the files we check earlier).

But skip the parentless TODOs are those one cannot be tracked (no file to check).

=item *

Finally, the todo hash will be converted into a markdown TODO list and depending on the given options will update the I<TODO.md> or print it to STDOUT.

By default I<todo> will update I<TODO.md> and only show pending tasks to STDOUT.

=back

=head1 IGNORE

If you add a I<.todoignore> file at the root of the directory to analyse, I<todo> will skip matching path(s).

=head1 AUTHOR

4zv4l I<4zv4l@protonmail.com>
